autoload/lsp/lspClient.vim

```diff
@@ -23,50 +24,58 @@ function! s:_on_lsp_stdout(id, data, event)
         endif
 
         while 1
-            if l:client.stdout.next_token == s:lsp_token_type_contentlength
-                let l:new_line_index = stridx(l:client.stdout.buffer, "\r\n")
-                if l:new_line_index >= 0
-                    let l:content_length_str = l:client.stdout.buffer[:l:new_line_index - 1]
-                    let l:client.stdout.buffer = l:client.stdout.buffer[l:new_line_index + 2:]
-                    let l:client.stdout.current_content_length = str2nr(split(l:content_length_str, ':')[1], 10)
-                    let l:client.stdout.next_token = s:lsp_token_type_contenttype
-                    continue
-                else
-                    " wait for next buffer to arrive
-                    break
-                endif
-            elseif l:client.stdout.next_token == s:lsp_token_type_contenttype
-                let l:new_line_index = stridx(l:client.stdout.buffer, "\r\n")
-                if l:new_line_index >= 0
-                    let l:content_type_str = l:client.stdout.buffer[:l:new_line_index - 1]
-                    let l:client.stdout.buffer = l:client.stdout.buffer[l:new_line_index + 4:]
-                    let l:client.stdout.current_content_type = l:content_type_str
-                    let l:client.stdout.next_token = s:lsp_token_type_message
+            if l:client.stdout.content_length == -1         " if content-length is -1 we haven't parsed the headers
+                " wait for all the headers to arrive
+                let l:header_end_index = stridx(l:client.stdout.buffer, "\r\n\r\n")
+                if l:header_end_index >= 0
+                    for l:header in split(l:client.stdout.buffer[:l:header_end_index - 1], "\r\n")
+                        let l:header_key_value_seperator = stridx(l:header, ":")
+                        let l:header_key = s:trim(l:header[:l:header_key_value_seperator - 1])
+                        let l:header_value = s:trim(l:header[l:header_key_value_seperator + 1:])
+                        if l:header_key ==? 'Content-Length'
+                            let l:client.stdout.content_length = str2nr(l:header_value, 10)
+                        endif
+                        let l:client.stdout.headers[l:header_key] = s:trim(l:header_value)

```

Comment:
currently I'm not exposing the headers to `on_notification`. If someone else requests for it will add it then.
header values are always string. 
http header can contains multiple keys but for now will leave it as { string: string } might need to change this to { string:  list of string } when we finally expose it to `on_notification`